package/MDAnalysis/core/selection.py

```diff
@@ -805,17 +775,31 @@ def __init__(self, parser, tokens):
         uppers = []  # upper limit on any range
         lowers = []  # lower limit on any range
 
-        for val in values:
+        _values = []
+        while values:
+            v = values.pop(0)
+            if v in ("to", ":", "-"):
+                try:
+                    _values[-1] = f"{_values[-1]} {v} {values.pop(0)}"
+                except IndexError:
+                    given = f"{' '.join(_values)} {v} {' '.join(values)}"
+                    raise ValueError(f"Invalid expression given: {given}")
+            else:
+                _values.append(v)
+
+        for val in _values:
             try:
-                lower = int(val)
+                lower = self.dtype(val)
                 upper = None
             except ValueError:
                 # check if in appropriate format 'lower:upper' or 'lower-upper'
-                selrange = re.match("(\d+)[:-](\d+)", val)
+                selrange = re.match(self.pattern, val)
                 if not selrange:
                     errmsg = f"Failed to parse number: {val}"
                     raise ValueError(errmsg) from None
-                lower, upper = np.int64(selrange.groups())
+                lower, upper = map(self.dtype, selrange.groups())
+                if lower > upper:
+                    lower, upper = upper, lower
```

Comment:
Should we really correct input? Wouldn't it be better to raise an error or return an empty selection? I would tend towards raising a ValueError (or SelectionError??).

If we did it before then now would be a good time to change it.

Reply:
current develop:

```python
>>> import MDAnalysis as mda
>>> from MDAnalysis.tests.datafiles import PDB
>>> u = mda.Universe(PDB)
>>> u.select_atoms("resid 5:3").residues.resids
array([3, 5])
>>> u.select_atoms("resid 3:5").residues.resids
array([3, 4, 5])
```

The selection is inclusive on both sides so it wouldn't necessarily return an empty selection. I guess it comes down to which behaviour is more expected (returning 3, 4, 5 for "5:3" or just 3, 5)

Reply:
I did not consider the inclusivity of both ends although I would still argue (from Python ranges) that `5:3` should only return something with a negative step `5:3:-1` and then it should reverse order. Which does not make much sense here.

What does VMD/PyMOL/... do for such a selection? (@IAlibay @RMeli do you know? Any suggestions?)

I think we can ultimately do whatever we think makes most sense but compatibility with established programs might be useful.


Parenthetically: should we drop `:` as a range separator so that Python users don't think of it as a range?

Reply:
So for pymol, doing the selection "sel test, resid 42:45" will return you residues 42, 43, 44, 45. But doing "sel test, resid 45:42" will return nothing.

For VMD the closest selection I can think of would be "resid 42 to 45" and "resid 4 to 1" which match pymol's behaviour.

Personally, I think it would make sense to match this behaviour, i.e. ranges should always be inclusive & "positive". Whilst I could probably think of a complex scenario where having the range going in the "negative" direction might be useful, I can't see how it would be too much to ask users to keep to these rules.

On a separate note, it would be really cool to have a third argument to the range selection for a `step` value :)

Reply:
Both VMD and Pymol select no atoms in the reverse order. VMD only uses "to", Pymol only uses "-" and ":", all are inclusive both ends. X-Plor also uses ":" inclusive I think. Given the difference in Python vs MD convention maybe it is best to drop ":"...

Reply:
Ah I forgot `ResidSelection` is its own thing. `molnums` is the original range selection and returns nothing:

```python
>>> np.unique(u.select_atoms('molnum 0-5').molnums)
array([0, 1, 2, 3, 4, 5])
>>> np.unique(u.select_atoms('molnum 5-0').molnums)
array([], dtype=int64)
```

Reply:
Thanks for all the research, @IAlibay & @lilyminium , how about going forward with the following
- keep `:` as a separator because  it is established in a number of tools (CHARMM is also using it https://www.charmm.org/charmm/documentation/by-version/c45b1/select.html), and we would potentially break existing code needlessly — but I can be convinced otherwise (i.e., to drop it for 2.0.0 as part of a revamping of the selection language)
- make reversed ranges return empty selections (as do many other tools) — state that behavior changed (i.e., ranges have to be monotonically increasing, lower ≤ upper, endpoints are inclusive)
- leave steps on ranges etc for future enhancements